contracts-periphery/test/Gas.t.sol

```diff
@@ -0,0 +1,423 @@
+
+  UmbraBatchSend.SendEth[] sendEth;
+  UmbraBatchSend.SendToken[] sendToken;
+
+
+  // Setup params for each function
+  enum Send {ETH, TOKEN, BOTH}
+  address payable[] addrs;
+  uint256 valueAmount; // {value: valueAmount} when calling Umbra
+  uint256 numOfAddrs;
+  uint256 etherAmount; // Ether amount to be sent per address
+  uint256 tokenAmount; // Token amount to be sent per address
+
+  function setUp() virtual public {
+    // Deploy Umbra at an arbitrary address, then place the resulting bytecode at the same address as the production deploys.
+    vm.etch(umbra, (deployCode("test/utils/Umbra.json", bytes(abi.encode(0, address(this), address(this))))).code);
+    router = new UmbraBatchSend(IUmbra(address(umbra)));
+    token = new MockERC20("Test","TT", 18);
+    token.mint(address(this), 1e7 ether);
+    deal(address(this), 1e5 ether);
+
+    ethBalance = address(this).balance;
+    tokenBalance = token.balanceOf(address(this));
+  }
+
+  function testPostSetupState() public {
+    uint256 currentToll = UmbraToll(umbra).toll();
+    assertEq(toll, currentToll);
+  }
+
+  function addParams(Send _type, uint256 numOfAddrs, uint256 etherAmount, uint256 tokenAmount) public {
+    token.approve(address(router), tokenBalance);
+    // Create a list of addresses
+    for(uint256 i = 0; i < numOfAddrs; i++) {
+      addrs.push(payable(address(uint160(uint(keccak256(abi.encode(i)))))));
+    }
+
+    if (_type == Send.ETH) {
+      for (uint256 i = 0; i < numOfAddrs; i ++) {
+        valueAmount += etherAmount + toll;
+        sendEth.push(UmbraBatchSend.SendEth(addrs[i], etherAmount, pkx, ciphertext));
+      }
+
+    } else if (_type == Send.TOKEN) {
+      for (uint256 i = 0; i < numOfAddrs; i ++) {
+        valueAmount += toll;
+        sendToken.push(UmbraBatchSend.SendToken(addrs[i], address(token), tokenAmount, pkx, ciphertext));
+      }
+
+    } else {
+        for (uint256 i = 0; i < numOfAddrs; i ++) {
+        valueAmount += etherAmount + toll * 2;
+        sendEth.push(UmbraBatchSend.SendEth(addrs[i], etherAmount, pkx, ciphertext));
+        sendToken.push(UmbraBatchSend.SendToken(addrs[i], address(token), tokenAmount, pkx, ciphertext));
+        }
+      }
+  }
+
+  // Send MAX balance
+  function addParams(Send _type, uint256 numOfAddrs) public {
+    token.approve(address(router), tokenBalance);
+
+    for(uint256 i = 0; i < numOfAddrs; i++) {
+      addrs.push(payable(address(uint160(uint(keccak256(abi.encode(i)))))));
+    }
+
+    if (_type == Send.ETH) {
+      for (uint256 i = 0; i < numOfAddrs; i ++) {
+        sendEth.push(UmbraBatchSend.SendEth(addrs[i], (ethBalance/numOfAddrs) - toll, pkx, ciphertext));
+      }
+      valueAmount = ethBalance;
+
+    } else if (_type == Send.TOKEN) {
+      for (uint256 i = 0; i < numOfAddrs; i ++) {
+        valueAmount += toll;
+        sendToken.push(UmbraBatchSend.SendToken(addrs[i], address(token), (tokenBalance/numOfAddrs), pkx, ciphertext));
```

Comment:
That's a good idea, let's add an assertion in `executeParams` that ensures `etherAmount` and `tokenAmount` are greater than zero

Reply:
Added a few `require`s in latest commit https://github.com/ScopeLift/umbra-protocol/pull/338/commits/634ca1fa74a8264033cfc92318c58f3d7ab3eaf9

Reply:
Why `require` instead of the standard `assert*` methods? It doesn't really matter much, just kinda seems out of place.

Also `etherAmount >= 0 && tokenAmount >= 0` is always true because the types are `uint256`, which are by definition always >= 0.

Reply:
Also why do we check `ethBalance > 0 && tokenBalance > 0` here? These are state variables defined only once in `setUp` so if anything they should be tested in the `testPostSetupState` test. It's a bit confusing to have a test take a few inputs, then immediately make assertions about state variables

Reply:
Not sure what I was thinking there, updated checks in https://github.com/ScopeLift/umbra-protocol/pull/338/commits/5a96dc632df0dbadcef957a5645e5542d0c361f8

Reply:
Wait, why did we get rid of the checks on `etherAmount` and `tokenAmount`? We want to ensure those are non-zero right?

Reply:
So I guess I could add `assert(etherAmount > 0)`, `assert(tokenAmount>0)`, and `assert(etherAmount > 0 && tokenAmount >0)` corresponding to each `Send _type`. Do you want me to do that?

Reply:
Yep I think keying off `_type` to know to assert seems like a good approach

Reply:
Added the checks https://github.com/ScopeLift/umbra-protocol/pull/338/commits/e46121b5609712b1248fba518766e48ffaa2858a

Reply:
We should use `assertEq` instead of solidity's `assert`. The former is a test helper that foundry recognizes and will display logs, the latter is a solidity method that reverts and consumes all gas, and isn't really used too often (since `require` and `revert()` don't consume all gas)

Reply:
You mean `assertTrue` in our case, right?

Reply:
Sounds good, will take note